src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -89,84 +133,82 @@ private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionCont
+                    // Move caret after the do statment's closing paren.
+                    var doStatementCaret = args.SubjectBuffer.CurrentSnapshot.GetPoint(((DoStatementSyntax)statementNode).CloseParenToken.Span.End);
+                    args.TextView.TryMoveCaretToAndEnsureVisible(doStatementCaret);
                     return;
-                }
-
-                if (currentNode.Parent == null)
-                {
+                case SyntaxKind.ForStatement:
+                    // `For` statements can have semicolon after initializer/declaration or after condition.
+                    // If caret is in initialer/declaration or condition, AND is inside other delimiters, complete statement
+                    // Otherwise, return without moving the caret.
+                    if (TryGetForStatementCaret(caret, statementNode, args, out var forStatementCaret) && isInsideDelimiters)
+                    {
+                        args.TextView.TryMoveCaretToAndEnsureVisible(forStatementCaret);
+                    }
+                    return;
+                case SyntaxKind.ExpressionStatement:
+                case SyntaxKind.GotoCaseStatement:
+                case SyntaxKind.LocalDeclarationStatement:
+                case SyntaxKind.ReturnStatement:
+                case SyntaxKind.ThrowStatement:
+                case SyntaxKind.FieldDeclaration:
+                    // These statement types end in a semicolon. 
+                    // if the original caret was inside any delimiters, `caret` will be after the outermost delimiter
+                    if (isInsideDelimiters)
+                    {
+                        args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(caret));
+                    }
+                    return;
+                default:
+                    // For all other statement types, don't complete statement.  Return without moving the caret.
                     return;
-                }
-
-                currentNode = currentNode.Parent;
-            }
-
-            // if the statement syntax itself requires a closing delimiter, verify it is there
-            if (StatementClosingDelimiterIsMissing(currentNode))
-            {
-                // Example: missing final `)` in `do { } while (x$$`
-                return;
             }
-
-            var semicolonPosition = GetSemicolonLocation(root, currentNode, caretPosition);
-
-            // Place cursor after the statement
-            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(semicolonPosition));
         }
 
-        private static int GetSemicolonLocation(SyntaxNode root, SyntaxNode currentNode, int caretPosition)
+        private bool TryGetForStatementCaret(SnapshotPoint originalCaret, SyntaxNode statementNode, TypeCharCommandArgs args, out SnapshotPoint forStatementCaret)
```

Comment:
Can the parameter be `ForStatementSyntax` instead of `SyntaxNode` since we're just casting it anyways?

Reply:
Yes, fixed it.